### PR TITLE
implement difficulty hardfork (no tests)

### PR DESCRIPTION
**Pull request by DavidVorick**
This is the long-awaited difficulty hardfork that has been holding back the v1.3.0 betas. Please do not merge until I've had time to write more tests and go over this with a fine toothed comb.

The design and algorithm are tested, simulated, and well thought through, including edge cases and adversarial situations. The design itself should be ready to rock, it's more a question of whether the implementation matches the design or if there's some egregious mistake somewhere.

### CODE REVIEW

File: modules/consensus/difficulty.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+	}
+
+	// For each value, first multiply by the decay, and then add in the new
+	// delta.
+	newTotalTime = (prevTotalTime * types.OakDifficultyDecayNum / types.OakDifficultyDecayDenom) + (int64(currentTimestamp) - int64(parentTimestamp))
+	newTotalTarget = prevTotalTarget.MulDifficulty(big.NewRat(types.OakDifficultyDecayNum, types.OakDifficultyDecayDenom)).AddDifficulties(targetOfCurrentBlock)
+
+	// Store the new total time and total target in the database at the
+	// appropriate id.
+	bytes := make([]byte, 72, 72)
```
Comment (lukechampine) [inline modules/consensus/difficulty.go]: you can omit the capacity if it's the same as the length

File: types/constants.go [hunk 2]
```diff
@@ -39,6 +45,11 @@ var (
 	// The GenesisID is used in many places. Calculating it once saves lots of
 	// redundant computation.
 	GenesisID BlockID
+
+	// Oak hardfork constants.
+	OakHardforkBlock     BlockHeight
+	OakDifficultyMaxRise = big.NewRat(1004, 1000)
+	OakDifficultyMaxDrop = big.NewRat(1000, 1004)
```
Comment (lukechampine) [inline types/constants.go]: do these belong in `types`? I thought previous hardfork values lived in `consensus`. I feel like they make more sense there.

File: modules/consensus/consensusdb.go [hunk 3]
```diff
@@ -62,6 +69,18 @@ var (
...
+	// FieldOakInit is a field in BucketOak that gets set to "true" after the
+	// oak initialiation process has completed.
+	FieldOakInit = []byte("OakInit")
+)
+
+var (
+	// ValueOakInit is the value that the oak init field is set to if the oak
+	// difficulty adjustment fields have been correctly intialized.
+	ValueOakInit = []byte("true")
+)
```
Comment (lukechampine) [inline modules/consensus/consensusdb.go]: why are these exported? (why are any of the bucket names exported?)

File: types/constants.go (hunk 2 above)
Comment (DavidVorick) [inline types/constants.go]: The existing difficulty adjustment values live in `types` already, using the names `MaxAdjustmentUp` and `MaxAdjustmentDown`.

Things belong in the `types` package if other, non-consensus packages might need to see the constants, because they can't necessarily import the consensus package (in practice they can).

I'll think about package placement a bit more, but I think the  types package is the right place for the hardfork height.

File: modules/consensus/consensusdb.go (hunk 3 above)
Comment (DavidVorick) [inline modules/consensus/consensusdb.go]: I am not sure, I don't think the bucket names need to be exported. But if we un-export them, that should happen in another PR. This PR has it because it's consistent.

File: types/constants.go (hunk 2 above)
Comment (lukechampine) [inline types/constants.go]: I guess it's consistent with functions like [`types.Tax`](https://github.com/NebulousLabs/Sia/blob/2a58b7e9a314173ae0d8390a66f94227feab194a/types/filecontracts.go#L121) including hardfork heights

Comment (DavidVorick) [inline types/constants.go]: I'm going to leave it in `types` because that's where all of the other consensus tuning parameters are. Makes sense to me to keep everything together. I'll clean up the comments at least though.
